src/workerd/api/tests/htmlrewriter-test.js

```diff
@@ -610,6 +610,157 @@ export const manualWriting2 = {
+
+    // This variation uses the JavaScript TransformStream, so we can
+    // initiate the read after doing the writes.
+    const promise = response.text();
+    strictEqual(
+      await promise,
+      `<html><b>such markup <i>much wow</i></b>bar</html>`
+    );
+  },
+};
+
+export const streamingReplacementReplace = {
+  async test() {
+    const { readable, writable } = new TransformStream();
+
+    const response = new HTMLRewriter()
+      .on('.dinosaur', {
+        async element(element) {
+          const dataStream = (await fetch('data:,goodbye world')).body;
+          element.replace(dataStream, { html: false });
+        },
+      })
+      .transform(new Response(readable));
+
+    const writer = writable.getWriter();
+    const encoder = new TextEncoder();
+
+    await writer.write(encoder.encode('<html>'));
+    await writer.write(
+      encoder.encode('<div class="dinosaur">hello world</div>')
+    );
+    await writer.write(encoder.encode('</html>'));
+    await writer.close();
+
+    // This variation uses the JavaScript TransformStream, so we can
+    // initiate the read after doing the writes.
+    const promise = response.text();
+    strictEqual(await promise, `<html>goodbye world</html>`);
+  },
+};
+
+export const streamingReplacementMultiple = {
+  async test() {
+    const { readable, writable } = new TransformStream();
+
+    const response = new HTMLRewriter()
+      .on('*', {
+        async element(element) {
+          element.prepend(await fetch('data:,alpha%20'));
+          element.append(await fetch('data:,%20gamma'));
+        },
+      })
+      .transform(new Response(readable));
+
+    const writer = writable.getWriter();
+    const encoder = new TextEncoder();
+
+    await writer.write(encoder.encode('<html>'));
+    await writer.write(encoder.encode('beta'));
+    await writer.write(encoder.encode('</html>'));
+    await writer.close();
+
+    // This variation uses the JavaScript TransformStream, so we can
+    // initiate the read after doing the writes.
+    const promise = response.text();
+    strictEqual(await promise, `<html>alpha beta gamma</html>`);
+  },
+};
+
+export const streamingReplacementBadUTF8 = {
+  async test() {
+    const { readable, writable } = new TransformStream();
+
+    const response = new HTMLRewriter()
+      .on('*', {
+        async element(element) {
+          element.prepend(await fetch('data:,garbage%e2%28%a1'));
+        },
+      })
+      .transform(new Response(readable));
+
+    const writer = writable.getWriter();
+    const encoder = new TextEncoder();
+
+    await writer.write(encoder.encode('<html>'));
+    await writer.write(encoder.encode('bar'));
+    await writer.write(encoder.encode('</html>'));
+    await writer.close();
+
+    // This variation uses the JavaScript TransformStream, so we can
+    // initiate the read after doing the writes.
+    // TODO(someday): Is it a good idea to pass on the specific message from lol_html? How can we do that?
```

Comment:
I think we might get a friendly message for free by returning `getLastError()` from our `lol_html_streaming_sink_write_utf8_chunk()` call sites.

Reply:
This works. Neat!